Refactor: Complete Script Rewrite

Mycroft-Studios opened the pull request:
**Describe Pull request**
A complete Rewrite of the script, into a modular based system, benefits of the rewrite include:
- Much easier to read
- Fully documented code
- Easy to follow each part
- Security improvements
- Performance improvements
   >  from my testing, it goes to 0.02ms max, and that is only for a single tick while dropping off the passenger, the rest of the time it will stay at 0.00ms
- Better coding practices
- Removes almost all threads
- Removes unused Code

**Questions (please complete the following information):**
- Have you personally loaded this code into an updated qbcore project and checked all it's functionality? Yes
- Does your code fit the style guidelines? Yes
- Does your PR fit the contribution guidelines? Yes

Mycroft-Studios committed:
Refactor: Complete Script Rewrite

A complete Rewrite of the script, into a OOP based system,
benefits of the rewrite include:
- Much easier to read
- Fully documented code
- Easy to follow each part
- Security improvements
- Performance improvements (from my testing, it goes to 0.02ms max, and that is only for a single tick while dropping off the passenger)
- Better coding practices
- Removes all threads

Mycroft-Studios committed:
tweak: slight performance improvements

credit to Mutt for noticing this :)

Jamie9192 commented:
Nicely structured and documented

Ive only scanned through the server lua - a performance improvement would be:

`Server.object = exports['qb-core']:GetCoreObject()` removed

`local GetPlayerObject = exports['qb-core']:GetCoreObject().Functions.GetPlayer` caching the single function, instead of the full core.

and call this function to get the players object :)

Z3rio commented on client/modules/npc.lua:
```diff
@@ -0,0 +1,258 @@
+local SetEntityAsMissionEntity = SetEntityAsMissionEntity
+local TaskGoStraightToCoord = TaskGoStraightToCoord
+local SetPedAsNoLongerNeeded = SetPedAsNoLongerNeeded
+local TaskEnterVehicle = TaskEnterVehicle
+local ClearPedTasksImmediately = ClearPedTasksImmediately
+local TaskEnterVehicle = TaskEnterVehicle
```
TaskEnterVehicle is duplicated

Z3rio replied:
```suggestion
local TaskEnterVehicle = TaskEnterVehicle
local ClearPedTasksImmediately = ClearPedTasksImmediately
```

Z3rio commented on client/modules/bus.lua:
```diff
@@ -0,0 +1,237 @@
+
+            exports["qb-core"]:HideText() -- Hide the text on the screen.
+        end
+    else 
```
Lint report:
client/modules/bus.lua:193:9: line contains trailing whitespace

Z3rio replied:
```suggestion
    else
```

Z3rio commented on fxmanifest.lua:
```diff
@@ -18,7 +18,8 @@ client_scripts {
     '@PolyZone/EntityZone.lua',
     '@PolyZone/CircleZone.lua',
     '@PolyZone/ComboZone.lua',
```
Only `CircleZone` is used, no?
I see no usage of `EntityZone`, `BoxZone`, `ComboZone` atleast

Mycroft-Studios replied:
maybe?
it has been a good while since i have used PolyZone, so didnt want to break it lol

Z3rio replied:
Yeah dont really know personally either, if the CircleZone perhaps depends on one of the other imported files.
If not though, then they should probably be removed to improve performance / overhead, no?

Mycroft-Studios replied:
> then they should probably be removed to improve performance / overhead, no?

indeed. will do some testing and see if the others are required :)

Z3rio reviewed (changes requested):
Just nitpicking.
Great PR mate.

Mycroft-Studios committed:
tweak(client/modules/npc): duplicated function reference

Co-authored-by: Zerio <neoboij@gmail.com>

Mycroft-Studios committed:
tweak(client/modules/bus): trailing whitespace

Co-authored-by: Zerio <neoboij@gmail.com>

Mycroft-Studios committed:
tweak(server): dont import full object for one function

Mycroft-Studios commented:
> Nicely structured and documented
> 
> Ive only scanned through the server lua - a performance improvement would be:
> 
> `Server.object = exports['qb-core']:GetCoreObject()` removed
> 
> `local GetPlayerObject = exports['qb-core']:GetCoreObject().Functions.GetPlayer` caching the single function, instead of the full core.
> 
> and call this function to get the players object :)

Fixed :)

Mycroft-Studios committed:
refactor: Remove OOP usage

Mycroft-Studios committed:
tweak(client/main): change onResourceStart to CreateThread

while this might sound like a weird change, onResourceStart does not seem to be 100% reliable in current builds
